src/platform_impl/linux/x11/mod.rs

```diff
@@ -453,61 +379,61 @@ impl<T: 'static> EventLoopProxy<T> {
 
 struct DeviceInfo<'a> {
     xconn: &'a XConnection,
-    info: *const ffi::XIDeviceInfo,
-    count: usize,
+    _reply: XcbBox<ffi::xcb_input_xi_query_device_reply_t>,
+    iter: ffi::xcb_input_xi_device_info_iterator_t,
 }
 
 impl<'a> DeviceInfo<'a> {
-    fn get(xconn: &'a XConnection, device: c_int) -> Option<Self> {
+    fn get(xconn: &'a XConnection, device: ffi::xcb_input_device_id_t) -> Option<Self> {
         unsafe {
-            let mut count = 0;
-            let info = (xconn.xinput2.XIQueryDevice)(xconn.display, device, &mut count);
-            xconn.check_errors().ok()?;
-
-            if info.is_null() || count == 0 {
-                None
-            } else {
-                Some(DeviceInfo {
-                    xconn,
-                    info,
-                    count: count as usize,
-                })
-            }
+            let mut err = ptr::null_mut();
+            let reply = xconn.xinput.xcb_input_xi_query_device_reply(
+                xconn.c,
+                xconn.xinput.xcb_input_xi_query_device(xconn.c, device),
+                &mut err,
+            );
+            let reply = match xconn.check(reply, err) {
+                Ok(i) => i,
+                Err(e) => {
+                    log::error!("Could not query device data: {}", e);
+                    return None;
+                }
+            };
+            let iter = xconn
+                .xinput
+                .xcb_input_xi_query_device_infos_iterator(&*reply);
+            Some(DeviceInfo {
+                xconn,
+                _reply: reply,
+                iter,
+            })
         }
     }
 }
 
-impl<'a> Drop for DeviceInfo<'a> {
-    fn drop(&mut self) {
-        assert!(!self.info.is_null());
-        unsafe { (self.xconn.xinput2.XIFreeDeviceInfo)(self.info as *mut _) };
-    }
-}
+impl<'a> Iterator for DeviceInfo<'a> {
+    type Item = *mut ffi::xcb_input_xi_device_info_t;
 
-impl<'a> Deref for DeviceInfo<'a> {
-    type Target = [ffi::XIDeviceInfo];
-    fn deref(&self) -> &Self::Target {
-        unsafe { slice::from_raw_parts(self.info, self.count) }
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.iter.rem > 0 {
+            let data = self.iter.data;
+            unsafe {
+                self.xconn
+                    .xinput
+                    .xcb_input_xi_device_info_next(&mut self.iter);
+            }
+            Some(data)
+        } else {
+            None
+        }
     }
 }
 
 #[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
-pub struct WindowId(ffi::Window);
-
-impl WindowId {
-    pub unsafe fn dummy() -> Self {
-        WindowId(0)
-    }
-}
+pub struct WindowId(ffi::xcb_window_t);
```

Comment:
I tried to test this but this causes a compiler error:
```
src/platform_impl/linux/mod.rs:151:43
    |
151 |         return WindowId::X(x11::WindowId::dummy());
    |                                           ^^^^^ function or associated item not found in `platform_impl::platform::x11::WindowId`
```

Reply:
Note that the error is also reported by GitHub Actions

Reply:
Weird it compiled fine for me and warned that this method is unused.

Reply:
Ok this function is only used if the Wayland feature is disabled. Please test with the Wayland feature enabled for now. I'll fix the problem next week.

Reply:
Fixed